### ISSUE TITLE
Vero 4K support

### ISSUE THREAD

**Issue by TheHacker66** [irtimmer/moonlight-embedded#573]
Please provide the following info.

**_NVidia Geforce Experience version:_** 3.9.0.97
**Moonlight Embedded version:** 2.4.3
**Moonlight Embedded source:** _compiled from source_
**Moonlight Embedded running on:** _Vero 4K_
**Moonlight Embedded running on distribution:** _OSMC_

**Verbose output `-verbose` of Moonlight Embedded:**

> Searching for server...
> Connect to 192.168.178.20...
> No mapping available for meson-ir (19000000000000000000000000000000) on /dev/input/event2
> No mapping available for gpio_keypad (10000000010000000100000000010000) on /dev/input/event0
> No mapping available for cec_input (100000008e1b0000ec0c000001000000) on /dev/input/event1
> Initializing platform...done
> Resolving host name...done
> Starting RTSP handshake...done
> Initializing control stream...done
> Initializing video stream...done
> Initializing audio stream...done
> Initializing input stream...done
> Starting control stream...done
> OPEN es DEVICE
> codec_init amstream version : 2.0
> [codec_init 732]type =1
> [codec_init 740]
> Starting video stream...done
> ALSA lib pcm_direct.c:896:(snd1_pcm_direct_initialize_slave) slave plugin does not support mmap interleaved or mmap noninterleaved access
> ALSA lib pcm_dmix.c:1034:(snd_pcm_dmix_open) unable to initialize slave
> Starting audio stream...Alsa error code -22
> Audio stream start failed: -1
> Stopping video stream...ENet wait interrupted
> Control stream connection failed
> Loss Stats: Transaction failed: 11
> [codec_close 851]
> [codec_close_cntl]video codec close return=0!
> done
> Stopping control stream...done
> Cleaning up input stream...done
> Cleaning up audio stream...done
> Cleaning up video stream...done
> Cleaning up control stream...done
> Cleaning up platform...done
> EVIOCGRAB failed with error 16
> EVIOCGRAB failed with error 16

**Comment by TheHacker66** [comment]
I'm bumping this and adding more logs:

https://gist.github.com/antsu/b426d72a35e7b32fbad64f05552225ba

Hope that someone will look into this sooner or later ;)

**Comment by cgutman** [comment]
Can you try specifying the target audio device manually?

If that fails, try building with PulseAudio to see if the PA backend works.

**Comment by TheHacker66** [comment]
Hi @cgutman!

Thanks for the suggestions, specifying the audio works (kinda), i can hear crackling sounds coming from the tv but the video is still black.

Something interesting coming from the verbose output:

`[codec_init 732]type =1
[codec_init 740]
Starting video stream...done
Starting audio stream...done
Starting input stream...done
EVIOCGRAB failed with error 16
EVIOCGRAB failed with error 16
Alsa error from writei: -32
Returning RTP packet queued for too long
Received OOS audio data (expected 28132, but got 28133)Waiting for IDR frame
Alsa shortm write, write 48 frames
Alsa shortm write, write 64 frames
Alsa shortm write, write 192 frames
Alsa shortm write, write 64 frames
Alsa error from writei: -32
Unrecoverable frame 197: 16+2=18 received < 21 needed
Network dropped an entire frame
codec_write error: ffffffff
[codec_close 851]
[codec_close_cntl]video codec close return=0!
OPEN es DEVICE
[codec_init 732]type =1
[codec_init 740]
[codec_reset:1199]ret=0
Requesting IDR frame on behalf of DR
Waiting for IDR frame
Waiting for IDR frame
Alsa error from writei: -32
...`

Hope this might give some clues!

**Comment by cgutman** [comment]
What exact parameters are you passing to stream? What GPU is in your PC?

You might try adding `-codec h264`. I found that some Amlogic chips don't like the default h.265 stream that GeForce Experience produces.

I implemented a workaround in moonlight-common-c for the most recent GFE version, which supports specifying the number of reference frames. https://github.com/moonlight-stream/moonlight-common-c/commit/f6ae7fc2b4d2c0c5d2a8cb8ebc55f4157156c9f1

**Comment by TheHacker66** [comment]
The last time I simply ran `moonlight stream -audio hw:0,0 -platform aml`

I merged the recent commits, I'll compile, test using your suggestions and report back tonight. Thanks!

**Comment by TheHacker66** [comment]
Unfortunately specifying the h264 codec didn't change much, i got less errors in console but the screen is still black.

**Comment by TheHacker66** [comment]
OSMC and Vero 4k were updated to Stretch. Unfortunately the black screen issue still remains.

**Comment by samnazarko** [comment]
Can you try run strace and see if it gives us anything useful?

**Comment by TheHacker66** [comment]
Yup, here's the log: https://paste.osmc.tv/isavozohoz

**Comment by danielfmo** [comment]
Hello,

I belibe this is a kodi and amlogic problem and not with moonlight.

Please take a look here:
https://forum.odroid.com/viewtopic.php?f=91&t=15456&p=171714&hilit=%2Fsys%2Fclass%2Fvideo%2Fdisable_video#p171714

try to set /sys/class/video/disable_video to 0 before starting moonlight

**Comment by samnazarko** [comment]
Good find @danielfmo . Let us know how it goes @TheHacker66 .
I can patch Kodi to reset this on exit; or you can set it in a startup script in Moonlight.

**Comment by TheHacker66** [comment]
Well guys, unfortunately setting that flag before running moonlight didn't change anything. I can run another strace with the flag set to 0 if needed.

So close yet so far..

**Comment by samnazarko** [comment]
@TheHacker66 I've sent you a PM with a kernel with amlvideo debugging on. It may give us some clues if the decoder isn't happy.

**Comment by TheHacker66** [comment]
I won't be at home till Saturday, so I'll test it once I get back.

**Comment by samnazarko** [comment]
Cool. I found why HEVC wasn't tolerated by Moonlight (buf margin needs a bump). This also fixed a couple of playback issues.

Sent from a mobile device, so please excuse the brevity of this reply

On 14 Feb 2018, at 19:30, TheHacker66 <notifications@github.com<mailto:notifications@github.com>> wrote:


I won't be at home till Saturday, so I'll test it once I get back.

-

**Comment by TheHacker66** [comment]
Hi, sorry for not replying earlier.

I installed the kernel and rebooted. Unfortunately there seems to be an incompatibility with the latest GFE version because moonlight complains about an udp port which is open in the firewall and that was working before. The version I have installed is 3.12.0.84

I tried uninstalling that and installing the version available on nvidia's site (3.12.0.79) to no avail. Deleting firewall rules (that were recreated after the install) didn't change anything.

I'll try to reinstall a even older version to see if the problem goes away.

**Comment by TheHacker66** [comment]
Ok, I reinstalled the version I had when I started this thread (3.9.0.97) and moonlight started working again.

Here's the dmesg log with disable_video set to 1:
https://paste.osmc.tv/ezeludacig

This is the dmesg log with the flag set to 0:
https://paste.osmc.tv/milavatana

**Comment by samnazarko** [comment]
Hopefully you're feeling better now.

Thanks for giving this a go. 

>[  +0.010327] codec:vdec1 video changed to 0 x 0 0 fps clk->200MHZ

I think this is the problem. The video decoder doesn't seem to be set up correctly. 

I believe Moonlight uses libamcodec. Can you confirm if you compiled Moonlight using the libamcodec-dev-osmc package?

**Comment by TheHacker66** [comment]
Hi Sam, thanks.

I have indeed installed vero3-libamcodec-dev-osmc. I modified the `cmake/FindAmlogic.cmake` file to point to `/usr/osmc/lib/` where libamcodec.so is located.

It is correct?

**Comment by TheHacker66** [comment]
Here's the output from moonlight:

http://paste.osmc.tv/poqayipobo.vbs

**Comment by samnazarko** [comment]
Yes -- that's one way to get the compile to use the OSMC versions of libamcodec.

>[  +0.007596] codec:vdec1 video changed to 0 x 0 0 fps clk->200MHZ

This still looks to be the problem. I'm not sure why the decoder is not getting hints properly; but I suspect something in Moonlight isn't setting up the amprivate structs properly.

**Comment by samnazarko** [comment]
@cgutman @danielfmo might have some more ideas.

I'll see if I can get access to an NVIDIA system to test this out with.

**Comment by TheHacker66** [comment]
You should still have remote access to my Vero 4k to fiddle with. I won't be at home tomorrow but I will be this weekend if you need me for tests.

**Comment by williamluke4** [comment]
@TheHacker66 Have you had any luck?

**Comment by TheHacker66** [comment]
Nope. I think we need @cgutman or @irtimmer help.

**Comment by samnazarko** [comment]
We're happy to get hardware out to resolve this.

**Comment by samnazarko** [comment]
Nothing yet unfortunately.
I can't think of anything we have significantly diverged with that would cause such issues. 

I'm happy to ship hardware to any developers. In the interim, I'm going to see if I can set up an NVIDIA system to look in to this myself.  I am away for a few weeks however.

**Comment by ZacWolf** [comment]
Anything on this yet?

**Comment by samnazarko** [comment]
Unfortunately nothing has moved this forward yet.

**Comment by maxanier** [comment]
Finally got around testing it myself on my new Vero4k+, but I got the same issue: Black screen, but some crackling audio.

(@samnazarko ) I am using the OSMC August update with the vero4k+ video improvements installed.

I have compiled moonlight-embedded from the master branch, but adjusted cmake/FindAmlogic.cmake to include the /opt/vero3/include and /usr/osmc/lib paths ([gist](https://gist.github.com/maxanier/3318bb1d00c73a935933f2272bd53df7)).

I tried 720,1080,4k as well as h264 and h265. Result is the same.
Logs look similar to TheHacker66 ones:
moonlight stream -platform aml -localaudio -codec h265 -1080 -verbose: [MoonlightLog](https://gist.github.com/maxanier/5849be0e6c3c71c649599b29b684ecb9) [dmesg](https://gist.github.com/maxanier/67169985310c44096aac27e8f019553d)

I tried to do some debugging, but the only thing I could figure out so far is that the `codec_write error: ffffffff` log message is caused by the codec  handle is temporarily unavailable (11 EAGAIN) in amcodec `codec_write` but this is probably normal behaviour. 
I do not really know where to continue looking since I can't really tell if the problem lies in moonlight (embedded/common-c) or in amcodec or somewhere else in the vero4k binaries.

**Comment by samnazarko** [comment]
Glad to hear you've received the new device.

Did you stop Kodi first? sudo systemctl stop mediacenter

**Comment by maxanier** [comment]
Yes

**Comment by samnazarko** [comment]
I understand. We need some assistance from the Moonlight guys here. Hopefully there is an opportunity to do this.

**Comment by cgutman** [comment]
The IDR frame warnings are a symptom of the problem. When `codec_write()` fails, the error triggers the IDR frame request to resync with the host (since video data was lost).

Can you see whether it's failing on the first `codec_write()` call or a later one? The AML decoder may expect to receive an entire frame at a time, rather than split up as we are currently submitting them. Can you try modifying the code to submit the entire frame at once like the SDL backend does? https://github.com/irtimmer/moonlight-embedded/blob/master/src/video/sdl.c#L66

From this Kodi fork, it looks like they are explicitly checking for EAGAIN and continuing to retry until it clears. You could also try that with or without the above change too. https://gitlab.com/ember-dev/kodi/blob/46dc687ff8c2ea711021b4b98a562eb9ce21a4d8/xbmc/cores/dvdplayer/DVDCodecs/Video/AMLCodec.cpp#L679

**Comment by maxanier** [comment]
Thank you very much for your response.

I implemented your first suggestion and it seems to fix the codec_write errors. Moonlight stops complaining.
However, the screen stays black. I still get the crackling (maybe slowed down) steam start sound (unless I use -localaudio).

[Updated aml.c](http://termbin.com/4b8z)
[Moonlight Verbose Output ](https://gist.github.com/maxanier/ba9511e79ba8611280d2b75e5ea18d94#file-moonlightlog)
[Dmesg](https://gist.github.com/maxanier/ba9511e79ba8611280d2b75e5ea18d94#file-dmesg)

I cannot tell much from it, but one thing that confuses me is a "init_decode_head_hw: (1920, 1088)", but I do not think this is related to our problem.

Just out of curiosity I logged the size of the frames passed to codec_write ([output](https://gist.githubusercontent.com/maxanier/ba9511e79ba8611280d2b75e5ea18d94/raw/e83ef0b0eaa35cef049a2f85471cb9d4046cd19b/write%2520log)). The output looks reasonable to me, about 30 logs per second, and the size starts large and gets smaller, probably because the transmitted frames is not changing (Steam start screen). But that are just guesses.

I also tried playing around with /sys/class/video/disable_video and more, but it does not help.

@samnazarko maybe something in osmc is preventing the decoded stream from actually being displayed. 
After stopping kodi, it displays the console output (from my rc.local), but after starting/stopping moonlight everything stays black until I restart kodi.

**Comment by samnazarko** [comment]
See https://github.com/osmc/libamcodec/blob/2017.10/amplayer/player/player_av.c#L1572 for how to set up decoder.

>I also tried playing around with /sys/class/video/disable_video and more, but it does not help.

You could try booting with Kodi disabled (systemctl disable mediacenter); in case Kodi or our watchdog are causing issues.

**Comment by maxanier** [comment]
Unfortunately, disabling kodi did not help.

I had a look at the code you linked, but could see anything interesting. I do not have any experience with video en/decoding, so I might have missed something relevant.

After that I had a look at https://github.com/osmc/libamcodec/blob/master/examples/esplayer.c which setups framebuffer and axis, but that did not help either.

Is there any way to get more debug log from the libs or kernel parts responsible for actually displaying what is passed to `codec_write`?

**Comment by samnazarko** [comment]
Sorry for the late reply on this.

I was able to modify the examples Makefile target and build simple_player.
With Big Buck Bunny I can initialise the VPU and see frames being decoded; but I only get a black screen. 

It's probably worth comparing the framebuffer state (sys/fb) when playing a video via Kodi vs trying to use Moonlight.

**Comment by maxanier** [comment]
Finally got around to try it again.
I updated the Vero to the December update, pulled all the latest upstream changes of moonlight-embedded and rebuild. But the issue remains.

I tried comparing the state of `/sys/class/graphics/fb0,1 /state,mode, ...` as well as `/sys/class/display/axis,mode`. But since I couldn't find any helpful documentation on the dozens of things I am not smarter than before. 
The only thing I noticed is that `/sys/class/graphics/axis` and `/sys/class/graphics/mode` are not updated if I stop kodi and start moonlight. The values of the last video playback remain, even if moonlight runs on a different resolution.

I am also not sure if it has any effect to directly modify those (e.g. `echo 2 > /sys/class/display/disable_video`). The values are update, but I never manged to see any difference on screen.

**Comment by TheHacker66** [comment]
Hey everyone, I also updated my fork and tried to no avail. Saw that some aml libraries were added to the cmake and fixed the paths for osmc but it didn't make a change.

hope @irtimmer or @cgutman have some other clue.

**Comment by maxanier** [comment]
I decided to give up on this. It feels like we are close, but without some experience with the Linux display system it is rather difficult to get it to work.
I am back to RPi3 for anything non 4k related, because neither moonlight nor hyperion work (well) on the vero.

**Comment by samnazarko** [comment]
Hyperion should work well in the next update.
Unfortunately I'm not sure how to proceed with Moonlight.

**Comment by EvgeniySpinov** [comment]
I'd like to join the battle for the moonlight + vero, cause I do not want to deal with 1080 on raspberry pi.

@samnazarko Is there any way to emulate Vero 4K+? I have only 1 device and it's used in living room and family would not be happy if I would take it for experiments. 

@maxanier Were you able to do super-verbose output by moonlight on connect (besides strace)? Is there such a mode? On your raspberry, you're also running OSMC? Upstream version and everything works? I also have Raspberry pi 3 where moonlight works without any hick ups. I have older version of the OSMC there though (Kodi 17.x).

**Comment by samnazarko** [comment]
>Is there any way to emulate Vero 4K+?

Not really unfortunately.

**Comment by cgutman** [comment]
Does the Vero 4K support the standard V4L2 stateful/stateless decoding APIs? If so, that's probably way to go, but Moonlight Embedded doesn't have a standard V4L2 backend yet. V4L2 is the way forward to get out of the proprietary vendor-specific API hell that is currently the embedded world.

The V4L2 stateful API is supported in upstream FFmpeg (`h264_v4l2m2m` and `hevc_v4l2m2m` decoders), but the V4L2 stateless API requires [out-of-tree patches](https://github.com/LibreELEC/LibreELEC.tv/tree/master/packages/multimedia/ffmpeg/patches/v4l2-request) (`v4l2request` hwaccel) while the kernel interface is still being completed. Which one you will need depends on how the Amlogic V4L2 kernel drivers are written.

If you want to test V4L2 on the Vero 4K, you can try building [moonlight-qt](https://github.com/moonlight-stream/moonlight-qt) with a V4L2-enabled FFmpeg build. Then you can test V4L2+EGL or V4L2+DRM rendering and see if performance is acceptable. Users on the RK3288/RK3399 with V4L2+EGL and V4L2+DRM on the Pi 4 for HEVC have reported good performance.

If V4L2 does work well, it shouldn't be too hard to drop in similar code in Moonlight Embedded to what Moonlight Qt. Moonlight Embedded doesn't have overlay support or windowed support (on most devices), so it's probably sufficient to just implement the V4L2+DRM path and omit V4L2+EGL.

**Comment by renatobellotti** [comment]
Has there been progress? :)

**Comment by samnazarko** [comment]
Unfortunately not.

**Comment by cgutman** [comment]
If the hardware offer still stands, I'd be happy to take a look.

Is there an FFmpeg build that supports hardware decoding on the Vero's current current media stack? FFmpeg would probably be easiest to integrate because Moonlight Qt supports it already.

Do you know what APIs the Kodi folks are using on this hardware? Is it V4L2+DRM or the old amcodec stuff?

**Comment by samnazarko** [comment]
Hi - we still use AMCodec on this device. It’s possible to support AMCodec via ffmpeg with some hacks but this hasn’t been tested. 

AML’s current focus is a V4L2 GStreamer stack. 

Happy to send some hardware.

**Comment by cgutman** [comment]
> Hi - we still use AMCodec on this device. It’s possible to support AMCodec via ffmpeg with some hacks but this hasn’t been tested.

So it sounds like our existing [amcodec support](https://github.com/moonlight-stream/moonlight-embedded/blob/master/src/video/aml.c) should be expected to work then. Hopefully it is something relatively simple.

> Happy to send some hardware.

OK, please reach out to cameron@moonlight-stream.org and we can get the shipping details sorted out.

**Comment by TheHacker66** [comment]
@cgutman I tried syncing and compiling my fork yesterday to see if updating to Debian 11 (which will be released soon on OSMC) changed anything but the same issue occurs.

Here's the fork https://github.com/TheHacker66/moonlight-embedded

Let me know if there's something I can do while you wait to receive the hardware.

Thanks for not abandoning OSMC support!

**Comment by samnazarko** [comment]
Thanks, I’ll get in touch with you shortly.

**Comment by samnazarko** [comment]
Hi -- I was travelling so it took a bit longer than expected to email you, but I've done so now and put a unit aside for you. Please let me know if you've received my email

**Comment by cgutman** [comment]
Yep, I got it and just replied. It got in my spam filter, so I didn't see it earlier.

**Comment by samnazarko** [comment]
This has shipped and I sent you tracking details. There’s currently a postal strike for two days in the UK, so I wouldn’t expect the package to move much until next week

**Comment by cgutman** [comment]
I worked on this for a few hours today trying various things and couldn't get anything working.

To see if it was Moonlight-specific, I tried the samples from the [OSMC Amcodec](https://github.com/osmc/libamcodec) repo and those exhibited the same symptoms (reporting successful decoding but not displaying anything).

The log messages from the kernel seem to be as expected:
```
[   78.416045] vdec_init, dev_name:ammvdec_h265, vdec_type=VDEC_TYPE_STREAM_PARSER  id = 0
[   78.416050] vdec_init set vfm decoder 00000000428cad09
[   78.416862] HEVC stbuf alloced at 00000000f2624b3e, secure = 0, size = 10485760
[   78.416977] pts_start, type=2
[   78.417534] video first pts = 0
[   78.417542] vdec_request_irq 00000000a9a08f5a, parser
[   78.418868] decoder_bmmu_box_alloc_box, tvp_flags = 0
[   78.419082] [0]dynamic_buf_num_margin=7
[   78.419088] [0]double_write_mode=0
[   78.419097] [0]h265:pts_unstable=0
[   78.419102] [0]h265: ver (201602101,0) decinfo: 1280x720 rate=1600
[   78.420499] [TEE], the fw (hevc_mmu_swap) will be loaded.
[   78.420602] hevc mux clock is 666666666 Hz
[   78.420608] hevc video changed to 3840 x 2160 60 fps clk->667MHZ
[   78.420898] ppmgr local_init
[   78.420911] di_receiver_event_fun: vframe provider reg ppmgr
[   78.421990] set run_early_proc_fun_flag to 1
[   78.422000] vpu_clkb = 0.
[   78.422001] DI: reg f
[   78.425832] ppmgr local_init
[   78.426125] vdec_init, vf_provider_name = decoder, b 0
[   78.426153] vdec->port_flag=0x102, port_flag=0x10b
[   78.451041] osmc videoenhancement: session context configured successfully
[   78.478751] [0]vh265 cached=16384  need_size=16384 speed= 60 ms
[   78.479573] [0]hevc->performance_profile 0
[   78.479587] [0]set pic_list_init_flag 1
[   78.481705] [0]init_buf_spec2 1280 736
[   78.481712] [0]set pic_list_init_flag to 2
[   78.481739] [0]set pic_list_init_flag to 3
[   78.481742] [vdec_kpi][vh265_isr_thread_fn] First I frame coming.
[   78.484248] [0]VH265: output first frame
[   78.484269] [vdec_kpi][post_video_frame] First I frame decoded.
[   78.484337] DI bypass all 1280x720-0x501000.
[   78.511486] cal_dur_from_pts inner driver dur1600 
[   78.533530] vframe_rate_uevent: sent uevent FRAME_RATE_HINT=1600
```

The `[   78.484248] [0]VH265: output first frame` line in particular seems to indicate that it did successfully output a frame to _somewhere_.

I also added a call to `codec_get_vdec_state()` before submitting each frame to confirm that the decoder is not reporting errors and it doesn't seem to be.

```
Received first video packet after 0 ms
Stream info: 0x0x60 | Errors: 0 | Status: 255
Stream info: 0x0x60 | Errors: 0 | Status: 255
Stream info: 1280x736x60 | Errors: 0 | Status: 255
```

The first two have no dimensions because the decoder hasn't processed enough of the bitstream yet, but then it does initialize (further indicate that the codec is indeed parsing the actual bitstream correctly and it's an output-related issue).

@samnazarko Is there something special that needs to be done to get amcodec output working? Since even the amcodec samples fail to play, I think this is probably related to Vero-specific kernel or amcodec changes.

I also tried moonlight-qt to see if maybe it could work more easily, but it looks like the Vero uses a legacy fbdev driver which is not something moonlight-qt supports. Even if it did support fbdev, it would not be able to do the performant zero-copy rendering that we can do with EGL or DRM renderers, so the performance likely wouldn't be good.

**Comment by samnazarko** [comment]
Hi,

The amcodec in that repository is quite old and we've a much more improved one which is installed in OSMC automatically. You can get the appropriate headers with ```sudo apt-get install libamcodec-dev-osmc```. That's just a newer library itself -- the examples are the same and unchanged.

**Comment by samnazarko** [comment]
According to @tanio99

>It might be related to the decoder type being used when initializing libamcodec. For HEVC playback one needs to set

codecParam.decoder_type = DECODER_TYPE_FRAME_MODE;

**Comment by cgutman** [comment]
> The amcodec in that repository is quite old and we've a much more improved one which is installed in OSMC automatically. You can get the appropriate headers with sudo apt-get install libamcodec-dev-osmc. That's just a newer library itself -- the examples are the same and unchanged.

Is the source for that available somewhere? I found it useful to look at the old one to try to figure out what the various functions and fields do.

> I assume you stopped Kodi first: sudo systemctl stop mediacenter

Yep

> It might be related to the decoder type being used when initializing libamcodec. For HEVC playback one needs to set `codecParam.decoder_type = DECODER_TYPE_FRAME_MODE`

When I set DECODER_TYPE_FRAME_MODE for HEVC, it seems to break even worse. `dmesg` fills up with `[0]cur lcu idx = 0, (total 920), set error_mark` messages and the error count as reported by `codec_get_vdec_state()` increases with each frame.

```
[  386.289232] the demux           clock on, ref cnt: 1
[  386.289469] the parser_top      clock on, ref cnt: 1
[  386.289474] the vdec            clock on, ref cnt: 1
[  386.289498] the clk_hevc_mux    clock on, ref cnt: 1
[  386.289508] hevc mux clock is 500000000 Hz
[  386.289623] vdec_create instance 000000004e80c985, total 1, PM: legacy
[  386.289722] no drmmode
[  386.289786] vdec_init, dev_name:ammvdec_h265, vdec_type=VDEC_TYPE_FRAME_BLOCK  id = 0
[  386.292733] decoder_bmmu_box_alloc_box, tvp_flags = 0
[  386.293169] [0]dynamic_buf_num_margin=7
[  386.293175] [0]double_write_mode=0
[  386.293183] [0]h265:pts_unstable=0
[  386.293189] [0]h265: ver (201602101,0) decinfo: 1280x720 rate=1600
[  386.293940] [TEE], the fw (hevc_mmu_swap) will be loaded.
[  386.294037] hevc mux clock is 666666666 Hz
[  386.294041] hevc video changed to 3840 x 2160 60 fps clk->667MHZ
[  386.294287] vfm map  created
[  386.294334] vf_reg_provider Error to notify receiver
[  386.294338] Error: vf_notify_receiver, fail to get receiver of provider vdec.h265.00
[  386.294341] Error: vf_notify_receiver, fail to get receiver of provider vdec.h265.00
[  386.294345] vdec_init, vf_provider_name = vdec.h265.00, b 0
[  386.320143] vdec->port_flag=0x102, port_flag=0x10b
[  386.324283] osmc videoenhancement: session context configured successfully
[  386.338968] [0]vh265 cached=16384  need_size=16384 speed= 40 ms
[  386.341230] [0]hevc->performance_profile 0
[  386.341253] [0]set pic_list_init_flag 1
[  386.343744] [0]init_buf_spec2 1280 736
[  386.343752] [0]set pic_list_init_flag to 2
[  386.343793] [0]set pic_list_init_flag to 3
[  386.343796] [vdec_kpi][vh265_isr_thread_fn] First I frame coming.
[  386.549268] [0]timeout_process decoder timeout
[  386.549277] [0]cur lcu idx = 151, (total 920), set error_mark
[  386.549318] [0]cur lcu idx = 151, (total 920), set error_mark
[  386.549813] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.549817] [0]VH265: output first frame
[  386.549837] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.550312] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.550754] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.550758] [0]VH265: output first frame
[  386.550808] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.551277] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.551283] [0]VH265: output first frame
[  386.551306] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.551753] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.551756] [0]VH265: output first frame
[  386.551773] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.552244] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.552665] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.552669] [0]VH265: output first frame
[  386.552684] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.553146] [0]cur lcu idx = 0, (total 920), set error_mark
[  386.553567] [0]cur lcu idx = 0, (total 920), set error_mark
```

> Additionally, the VFM map doesn't seem to be set in the logs.

Hmm, what's a VFM map?

> You should initialise the codec parameter structure fully, i.e:

This is what I'm doing now but no luck so far: https://github.com/moonlight-stream/moonlight-embedded/commit/87cad37958210dbcda2d2f14cccc99ad72449bb3

Do I need to do anything special with fb blanking for the Vero 4K? Right now Moonlight blanks both fb0 and fb1 on Amlogic devices. There are also things like `/sys/class/video/disable_video` (which defaults to 1) that maybe I need to mess with?

If you've got a simple amcodec sample that works on the Vero 4K, I can probably easily mimic what that does. I just couldn't get any of the ones from the old amcodec repo working at all.

**Comment by cgutman** [comment]
> As you see, without that map you will never ever get any frame out of the Vero.

On a freshly booted Vero (with Kodi disabled), it looks like vfm is set to `default { decoder(0) ppmgr(0) deinterlace(0) amlvideo(0) amvideo}` which works for us. Is that default initialization safe to rely on?

> You can then decide (e.g. based on that timestamp) when that frame should be displayed. You do that by calling the ReleaseFrame function, which tells the kernel that the given frame should be passed to the next module on the vfm map. The "amvideo" module is then doing all the magic to finally display that frame.

Aha! This was the missing part. Adding a pair of `VIDIOC_DQBUF` and `VIDIOC_QBUF` `ioctl()` calls makes the video display properly.

This code was previously used by other Amlogic decoders that apparently didn't need these ioctls to be invoked to display the output on screen. How do I tell programmatically if I'm dealing with a decoder that needs VIDIOC_DQBUF+VIDIOC_QBUF vs. one where `codec_write()` alone is enough? I don't want to break existing Amlogic devices with the Vero changes.

> Most often the reason is that the encoded frame had some errors in it.

Yep, the reason is obvious in hindsight. `DECODER_TYPE_FRAME_MODE` is telling the decoder we will give it full frames, but we call `codec_write()` multiple times with partial frames. `DECODER_TYPE_SINGLE_MODE` seems to work for me for both H.264 and HEVC without issues.

**Comment by tanio99** [comment]
> How do I tell programmatically if I'm dealing with a decoder that needs VIDIOC_DQBUF+VIDIOC_QBUF vs. one where codec_write() alone is enough?

It's not the decoder that needs the ioctl calls - it's the amlvideo module in the vfm map. If you remove it then you don't need the ioctl calls. But be aware - that's a setup which we've never used and tested. The setup would be:
```
echo add default decoder ppmgr deinterlace amvideo > /sys/class/vfm/map
```
If you want to check programmatically if you're running on a Vero, then you could use the output of
```
cat /proc/cpuinfo
```
and grep for "OSMC Vero".

**Comment by cgutman** [comment]
Okay, with https://github.com/moonlight-stream/moonlight-embedded/commit/b2192eda254fc55dff31ca0d6709b0a26661ef72 and my prior ALSA fixes, Moonlight now works pretty well on the Vero at 4K 30 FPS.

However, I am seeing significant video latency issues (independent of resolution) when streaming at 60 FPS. I assume this is a 59.94 Hz display vs 60 FPS stream issue which is causing those extra frames to accumulate in the display output pipeline. The accumulated delay can get up to 5+ _seconds_ when streaming for a while.

I'm not seeing extended delays from either `codec_write()` or `VIDIOC_DQBUF/VIDIOC_QBUF`, so it looks like the buffering is happening further along in the display pipeline, perhaps in `amvideo`.

Is there a way to either limit the number of frames buffered (preferably to 1 or 0), or request that amvideo only display the most recently queued frame? If neither of those are possible, is there a way I can synchronize with VBlank to dispatch frames in lock step with the rate they are being consumed by the output?

**Comment by tanio99** [comment]
> Okay, with https://github.com/moonlight-stream/moonlight-embedded/commit/b2192eda254fc55dff31ca0d6709b0a26661ef72 and my prior ALSA fixes, Moonlight now works pretty well on the Vero at 4K 30 FPS.

Great to hear that!

> I assume this is a 59.94 Hz display vs 60 FPS stream issue

Hm, if it's a fractional rate thing, then it could help to set the "fractional rate policy" like it is done in Kodi:
```
int fractional_rate = (res.fRefreshRate == floor(res.fRefreshRate)) ? 0 : 1;
CLog::Log(LOGDEBUG, "AMLUtils::aml_set_display_resolution setting frac_rate_policy to %d", fractional_rate);
SysfsUtils::SetInt("/sys/class/amhdmitx/amhdmitx0/frac_rate_policy", fractional_rate);
```
The buffering you've mentioned happens directly in amlvideo, right in the `ioctl(videoFd, VIDIOC_QBUF, &vbuf)` call, where the received frame is put into a queue. There's an interrupt service routine in amvideo running, which is being called 60 or 59.94 times a second depending on the refresh rate. The ISR will look at the queue in amlvideo then take a frame and display it.

If you want to sync with VBlank then you would need to use polling like it was done in AMLCodec.cpp (see `CAMLCodec::PollFrame()` function). But with kernel 4.9 that method became quite unreliable, since the thread wakeup time varies too much, and that was causing a lot of micro-stutter issues.

**Comment by TheChoconut** [comment]
It's great to see some progress on the AMLogic backend, I have been using hacks ever since to try and get this to work on CoreELEC... But this new code seems to work there too.

I've also noticed a delay at 60 FPS, which was happening on my end. Feeding pts into the decoder seems to fix this:
`codec_checkin_pts(&codecParam, decodeUnit->presentationTimeMs);`

I don't know why this works, but it seems to be fix the delay?

**Comment by cgutman** [comment]
@TheChoconut Can you submit your changes upstream? It looks like several of them should be acceptable as-is.

For the others (like 02d9f42761cc9cebec9754a68b363ccc9a8331b0), we can collaborate on a solution that works for all AML devices.

**Comment by samnazarko** [comment]
As always, I'm happy to provide hardware to participating developers to improvement outcomes for OSMC Vero owners.

Cheers

Sam

**Comment by tanio99** [comment]
These errors are new to me, need to investigate. Usually, errors like these are coming from the decoder and indicate issues with the stream.

**Comment by tanio99** [comment]
Ok, it seems there's something wrong with the stream. According to the H.264 spec:
```
When gaps_in_frame_num_value_allowed_flag is equal to 0, the bitstream shall not contain data that result in
frame_num not being equal to PrevRefFrameNum or ( PrevRefFrameNum + 1 ) % MaxFrameNum.
NOTE – When gaps_in_frame_num_value_allowed_flag is equal to 0 and frame_num is not equal to PrevRefFrameNum
and is not equal to ( PrevRefFrameNum + 1 ) % MaxFrameNum, the decoding process should infer an unintentional loss of
pictures.
```
there's a flag in the SEI indicating if frames may be missing or not. If that flag isn't set than the decoder assumes that frames are missing and it signals that situation with the error message above.

**Comment by tanio99** [comment]
No, there's unfortunately no way to get some metrics out of the decoder/video chain. I did some "performance tests" last year or so by adding some logs. The average decoding time for VP9 sources is about 17 ms per frame. For h.264 sources it varies depending on the frame type. I-Frames for instance can take up to 0.2 seconds, whereas other frames take only 2-5 ms. I did not measure h.265.

**Comment by samnazarko** [comment]
https://www.reddit.com/r/SteamDeck/comments/znxleo/gamestream_is_being_discontinued_meaning_the_end/

This doesn't look good.

**Comment by EvgeniySpinov** [comment]
Alternatively https://github.com/LizardByte/Sunshine can be used? I haven't tried it yet, but it seems that it's the way to go.

**Comment by EvgeniySpinov** [comment]
Was trying to build moonlight from master in order to test it with HEVC, however I'm getting:
```
RTSP port: 48010
Initializing platform...done
Resolving host name...done
Initializing audio stream...done
Starting RTSP handshake...Audio port: 48000
Video port: 47998
Control port: 47999
done
Initializing control stream...done
Initializing video stream...done
Initializing input stream...done
Starting control stream...done
Error: failed to open X display.
Starting video stream...Video stream start failed: -1
Stopping control stream...ENet peer acknowledged disconnection
done
Cleaning up input stream...done
Cleaning up video stream...done
Cleaning up control stream...done
Cleaning up audio stream...done
Cleaning up platform...done
EVIOCGRAB failed with error 16
```

Moonlight version: ```Moonlight Embedded 2.5.3-master-00c28f5```

Vero 4K, but OSMC version is not the most recent (kernel 3.14.29-160-osmc)

Is there any specific parameters I need to use or I should just upgrade?

**Comment by samnazarko** [comment]
You will need to be on a newer version of OSMC with the 4.9.269 kernel.

Your OSMC version will be very old (likely pre Aug 2021).

**Comment by EvgeniySpinov** [comment]
Thank you Sam. I'm tending not to update when everything works.

Right, I've updated to most recent version, got kernel 4.9.269-16-osmc, rebuilt moonlight, however the problem persists.

Also when I install ```sudo apt-get install rbp-userland-dev-osmc``` I get:

```
The following additional packages will be installed:
  rbp-userland-osmc
The following NEW packages will be installed:
  rbp-userland-dev-osmc rbp-userland-osmc
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/4934 kB of archives.
After this operation, 19.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
(Reading database ... 61451 files and directories currently installed.)
Preparing to unpack .../rbp-userland-osmc_3.8.0-2_armhf.deb ...
dpkg-divert: error: 'diversion of /etc/kernel-img.conf to /etc/kernel-img.conf.distributed by rbp-userland-osmc' clashes with 'diversion of /etc/kernel-img.conf to /etc/kernel-img.conf.distributed by vero3-userland-osmc'
dpkg: error processing archive /var/cache/apt/archives/rbp-userland-osmc_3.8.0-2_armhf.deb (--unpack):
 new rbp-userland-osmc package pre-installation script subprocess returned error exit status 2
Selecting previously unselected package rbp-userland-dev-osmc.
Preparing to unpack .../rbp-userland-dev-osmc_3.8.0-2_armhf.deb ...
Unpacking rbp-userland-dev-osmc (3.8.0-2) ...
Errors were encountered while processing:
 /var/cache/apt/archives/rbp-userland-osmc_3.8.0-2_armhf.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Not sure if it affects the build, but I was unable to overcome this. Otherwise error looks exactly the same as above.

Thoughts?

**Comment by samnazarko** [comment]
You shouldn’t have Raspberry Pi packages on this system (rbp-userland), what instructions did you follow?

Sam

**Comment by EvgeniySpinov** [comment]
Those from the project: https://github.com/moonlight-stream/moonlight-embedded/wiki/Full-details-for-Compilation-and-Installation-in-Raspbian-Or-OSMC

There is separate line for OSMC. Outdated?

**Comment by EvgeniySpinov** [comment]
Removed rbp-* packages and installed vero3-* instead. Now it looks like:
```
# dpkg -l | grep vero3-userland
ii  vero3-userland-dev-osmc              2.0.4                          armhf        Vero 3 userland package with headers
ii  vero3-userland-osmc                  2.0.4                          armhf        Vero 3 userland package
```

Rebuilt moonlight, however problem presists, i.e. nothing has changed. Not sure what could be done from here. Strace with start: https://pastebin.com/frj998n3

Need to mention that I'm executing this from SSH with media server shut down. Cannot execute it directly on device, as after shutting down media server I'm getting black screen without any output at all. Probably it's related.

**Comment by samnazarko** [comment]
Hi @EvgeniySpinov 

Starting from the command line should be fine. From your strace I can see Wayland and GBM libraries being loaded for windowing. On Vero 4K / 4K + we use GLES. 

I suspect that your binary has been built against the wrong windowing system; either because of some leftover Raspberry Pi packages or the presence of MESA on your system.

**Comment by EvgeniySpinov** [comment]
This is not exactly the case:

```
ldd /usr/local/bin/moonlight | grep vero
        libEGL.so.1 => /opt/vero3/lib/libEGL.so.1 (0xf5c0d000)
```

And in strace:

```
openat(AT_FDCWD, "/opt/vero3/lib/libEGL.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\3\0(\0\1\0\0\0\20|\2\0004\0\0\0tl\21\0\2\0\0\5"..., 512) = 512
fstat64(3, {st_mode=S_IFREG|0775, st_size=1387197, ...}) = 0
mmap2(NULL, 1177624, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0xf593f000
mprotect(0xf5a51000, 28672, PROT_NONE)  = 0
mmap2(0xf5a58000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x111000) = 0xf5a58000
mmap2(0xf5a5e000, 2072, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0xf5a5e000
mprotect(0xfff5e000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC|PROT_GROWSDOWN) = 0
close(3)                                = 0
openat(AT_FDCWD, "/opt/vero3/lib/libGLESv2.so.2", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\3\0(\0\1\0\0\0\20|\2\0004\0\0\0tl\21\0\2\0\0\5"..., 512) = 512
....
```

Unless you refer to another part where wayland and GBM are effectively used to determine X Display, since I haven't found such.

Probably there is env variable I need to set for moonlight to pick up display correctly?

**Comment by samnazarko** [comment]
It's good to see that Moonlight is linked against the correct libraries.

I haven't got an NVIDIA GPU to test this with; but I'm sure there are some compilation instructions somewhere in this repository for AMLogic devices.

**Comment by EvgeniySpinov** [comment]
Seems like I could use some help from @cgutman

Maybe there are some parameters I need to use to build moonlight in order to have it properly detect X display?

**Comment by EvgeniySpinov** [comment]
Probably this is the issue with display is about this: https://github.com/moonlight-stream/moonlight-embedded/wiki/Packages#additional-required-setup-steps-for-raspbian-bullseye

@samnazarko @tanio99 What is the way to have config.txt parameters in OSMC? It seems that recent version doesn't have config.txt anymore.

**Comment by samnazarko** [comment]
If you're referring to Raspberry Pi, we still have config.txt. 

However we now recommend editing config-user.txt, as we may overwrite config.txt as we see fit. This file is found on the /boot partition.

**Comment by EvgeniySpinov** [comment]
Actually I've meant Vero. You're saying I can create /boot/config-user.txt (file is not there), and put those configuration items in it and it will be picked up during boot? Or process is different for Vero?

**Comment by samnazarko** [comment]
No -- we don't have a config.txt on Vero and haven't in the past.
The commands suggested in that guide are for Raspberry Pi and aren't related - won't do anything on OSMC. 

From the instructions you've linked to, I can see that it enables a Moonlight repository, however the packages that it makes available will be for Raspberry Pi and not the Vero. I believe you'll need to compile it from source and specify AMLogic support when building so it works on the device.

**Comment by EvgeniySpinov** [comment]
What I've meant is this part of instructions:
![image](https://user-images.githubusercontent.com/10797822/224390423-56eb5bd0-b2db-41a4-88f9-f90fe5784fb5.png)

Is there a way to do this on Vero?

**Comment by samnazarko** [comment]
These instructions are Raspberry Pi specific, they’re not needed nor do they apply to Vero. But the packages the author is offering is built for Pi hardware not Vero.

**Comment by EvgeniySpinov** [comment]
Ok, given another try to moonlight-embedded. Updated and rebuilt. Tried without any third-party scripts, directly from console, in order to avoid any issues related to them.

Now version is: Moonlight Embedded 2.5.3-master-5bb47ce

Stopped media server manually and running via SSH:

```
$ moonlight stream -app 5 -verbose
Moonlight Embedded 2.5.3 (AML;SDL;X11;VDPAU;VAAPI;ALSA;PULSE;EMBEDDED)
Searching for server...
Connecting to 10.0.4.21...
GPU: , GFE: 3.23.0.74 (, 7.1.431.-1)
Platform SDL2 (software decoding)
The path /dev/dri/ cannot be opened or is not available
The path /dev/dri/ cannot be opened or is not available
Could not initialize SDL - No available video device
```

If I try with AML: 
```
$ moonlight stream -app 5 -verbose -platform aml
Moonlight Embedded 2.5.3 (AML;SDL;X11;VDPAU;VAAPI;ALSA;PULSE;EMBEDDED)
Searching for server...
Connecting to 10.0.4.21...
GPU: , GFE: 3.23.0.74 (, 7.1.431.-1)
Platform Unknown
Platform 'aml' not found
```

If somebody was successful launching it - please share.

**Comment by EvgeniySpinov** [comment]
Ok, I've ran some tests in different modes with Sunshine (0.18.4) + Moonlight and everything seems to work fine. I've tested resolutions from 720p till 2160p with rates from 8Mbps to 150Mbps. Wired network.

The only issue I can see if that HDR is not working. When I try to start HDR stream I get the log: https://paste.osmc.tv/akocujugax

Screen shows 1st frame and freezes till moonlight termination in around 10 seconds. Second attempt brings black screen.

HDR is supported on the host by GPU and monitor. And HDR is supported by Vero 4K+ and TV. Also Vero 4K+ can play HDR content from OSMC.

However when moonlight is started with HDR mode - TV is not entering HDR mode.

I've spent some time investigating issue I had with not being able to stream 4K and figured that I can stream 4K only when OSMC is configured to this resolution. Otherwise I get stream the same as initial resolution after boot. So my guess is that moonlight cannot change resolution on Vero 4K+ and switch on HDR as well. 

Would be glad to help with experiments in this area.

**Comment by samnazarko** [comment]
Hi

I’m currently travelling on business and behind on a few things.
I’ll make a note to follow up on this early next week.

Sam

From: EvgeniySpinov ***@***.***>
Reply to: moonlight-stream/moonlight-embedded ***@***.***>
Date: Tuesday, 11 April 2023 at 23:27
To: moonlight-stream/moonlight-embedded ***@***.***>
Cc: Sam Nazarko ***@***.***>, Mention ***@***.***>
Subject: Re: [moonlight-stream/moonlight-embedded] Vero 4K support (#573)


Ok, I've ran some tests in different modes with Sunshine (0.18.4) + Moonlight and everything seems to work fine. I've tested resolutions from 720p till 2160p with rates from 8Mbps to 150Mbps. Wired network.

The only issue I can see if that HDR is not working. When I try to start HDR stream I get the log: https://paste.osmc.tv/akocujugax

Screen shows 1st frame and freezes till moonlight termination in around 10 seconds. Second attempt brings black screen.

HDR is supported on the host by GPU and monitor. And HDR is supported by Vero 4K+ and TV. Also Vero 4K+ can play HDR content from OSMC.

However when moonlight is started with HDR mode - TV is not entering HDR mode.

I've spent some time investigating issue I had with not being able to stream 4K and figured that I can stream 4K only when OSMC is configured to this resolution. Otherwise I get stream the same as initial resolution after boot. So my guess is that moonlight cannot change resolution on Vero 4K+ and switch on HDR as well.

Would be glad to help with experiments in this area.

**Comment by cgutman** [comment]
> Server side (verbose, Software encoder enforced, NVENC doesn't support HDR on GTX 970)

Sunshine doesn't support HDR encoding with the software encoder. That's why you see all those "Warning: No compatible capture formats for this encoder" messages and Moonlight complains that no video was ever received.

**Comment by cgutman** [comment]
> if software encoder update to support HDR is planned?

Not at the moment. Color conversion on the CPU is already really slow even with 8bpc SDR frames, because it's currently single-threaded. HDR frames are also twice the size of SDR frames, which makes the performance problem even worse.

It also doesn't look like libswscale supports doing those linear gamma FP16 to Rec 2100 PQ color conversions that the we need to handle the HDR frames that Windows captures. I haven't looked really deeply though, so I may be wrong.

**Comment by EvgeniySpinov** [comment]
Thank you for detailed explanation. Probably I'm missing something, but to my knowledge software encoding could be multi-threaded (Sunshine has this setting as well), and overhead is coming from sync of frames provided by multiple streams, which is reduced by optimizations made within x264 encoder, used by ffmpeg library, which Sunshine utilizes. Am I getting it wrong? Or there are some specifics within Sunshine on output pipeline?

From my personal experience, I see no difference in latency between NVENC (x264) and Software encoder (x264). In both cases it's noticable at 4K@80Mbit/s comparing to native display. Reducing bitrate and/or resultion reduces latency for both of them. I'm using Ryzen 3600 - not a most advanced CPU. As well as 2 to 3 encoding threads (still have to find a sweet spot). CPU utilization is around 50-70% during game session, depending on game/scene/etc. More capable CPUs should reduce latency and utilization even further. 

To summarize, from my experience, so far, it seems that increasing 20-50% on load for encoder (this is average computational effort increase to encode 10bit HDR frame, even though the size of frame is 1.25 times bigger compared to 8bit SDR frame), should still keep software econding as a viable option. This is in theory, and numbers though are a bit different (in a way of reduction) than you've provided. If you can share more details on why it's 2 times the frame size and what latency it might introduce, even in theory, let's say in 1080p@60fps@40Mbit stream - that would be great.

As for libswscale, it seems that you're right. There is workaround though, through using zscale filter from z.lib, which supports high bit-depth and wide color gamut conversions, including conversions involving FP16 and Rec. 2100 PQ. Not sure how hard it would be to incorporate into Sunshine though.

**Comment by samnazarko** [comment]
Hi

I am back from my travel now.

I think everything’s been covered here but if I can help or if there’s anything that needs looking in to me do let me know

Many thanks,

Sam

From: EvgeniySpinov ***@***.***>
Reply to: moonlight-stream/moonlight-embedded ***@***.***>
Date: Tuesday, 18 April 2023 at 08:21
To: moonlight-stream/moonlight-embedded ***@***.***>
Cc: Sam Nazarko ***@***.***>, Mention ***@***.***>
Subject: Re: [moonlight-stream/moonlight-embedded] Vero 4K support (#573)


Thank you for detailed explanation. Probably I'm missing something, but to my knowledge software encoding could be multi-threaded (Sunshine has this setting as well), and overhead is coming from sync of frames provided by multiple streams, which is reduced by optimizations made within x264 encoder, used by ffmpeg library, which Sunshine utilizes. Am I getting it wrong? Or there are some specifics within Sunshine on output pipeline?

From my personal experience, I see no difference in latency between NVENC (x264) and Software encoder (x264). In both cases it's noticable at ***@***.***/s comparing to native display. Reducing bitrate and/or resultion reduces latency for both of them. I'm using Ryzen 3600 - not a most advanced CPU. As well as 2 to 3 encoding threads (still have to find a sweet spot). CPU utilization is around 50-70% during game session, depending on game/scene/etc. More capable CPUs should reduce latency and utilization even further.

To summarize, from my experience, so far, it seems that increasing 20-50% on load for encoder (this is average computational effort increase to encode 10bit HDR frame, even though the size of frame is 1.25 times bigger compared to 8bit SDR frame), should still keep software econding as a viable option. This is in theory, and numbers though are a bit different (in a way of reduction) than you've provided. If you can share more details on why it's 2 times the frame size and what latency it might introduce, even in theory, let's say in ***@***.***@40Mbit stream - that would be great.

As for libswscale, it seems that you're right. There is workaround though, through using zscale filter from z.lib, which supports high bit-depth and wide color gamut conversions, including conversions involving FP16 and Rec. 2100 PQ. Not sure how hard it would be to incorporate into Sunshine though.

**Comment by EvgeniySpinov** [comment]
Ok, seems like I've found 1 more issue:
- Sunshine 0.18.4, with NEVC enforced
- Moonlight from master, set to h264 codec

Unable to start streaming:
Logs from client (debug): https://paste.osmc.tv/sidasowupi
Logs from server (debug): https://paste.osmc.tv/ozamefumih.hs

Changing client codec to h265 or auto works without issues. I assume the issue is with those lines on Vero 4K+:
```
set_decoder_config
codec_h_ioctl_set failed,handle=25,cmd=111,paramter=0, t=ffffffff errno=19
codec_h_ioctl failed,handle=25,cmd=401053c2,subcmd=111, paramter=0, t=ffffffff errno=19
codec_init error: fefffff6
Starting video stream...Video stream start failed: -2
Stopping control stream...ENet peer acknowledged disconnection
done
```

i.e. seems that something wrong with codec setup on Vero 4K side, as stream is not even starting from server perspective.

**Comment by EvgeniySpinov** [comment]
Short update on H264 decoding: 
- Applying changes mentioned above from XBMC thread, I'm able to get video stream on Vero 4K+
- But it instantly starts to get increased latency with arythmetical progression and in around 10 seconds freezes completely. No errors in console (starting in debug mode, but probably there is some more). Something similar to what is described in this post: https://github.com/moonlight-stream/moonlight-embedded/issues/573#issuecomment-1308253584
- This seems like an issue with decoder pipeline setup or decoder settings, which is discussed above. However, I do not have enough knowledge to tackle this. Issue observed also with 1080p.

If anyone can help with this - would be great.

**Comment by TheHacker66** [comment]
Hi, I'm resuming the thread because I tried compiling from source but I still cannot see the video.

I assumed all the needed fixes were merged in master?

**Comment by antoyo** [comment]
Any update on this?

I also tried compiling from source and I cannot see the video.

I tried `echo 0 > /sys/class/video/disable_video`, but it didn't help.

**Comment by cgutman** [comment]
@samnazarko anything special we need for Vero V?

#884 logs see to show the codec is not consuming input quickly (at all?) - lots of `codec_write() timeout`

**Comment by antoyo** [comment]
I tried h264, h265 and av1. Same issue for all.

**Comment by samnazarko** [comment]
OK, thanks for the feedback.

**Comment by antoyo** [comment]
I made some tiny bit of progress.
By running these commands first, I see the first frame of the stream, then it goes black:
```
echo rm default > /sys/class/vfm/map
echo add default decoder ppmgr deinterlace amvideo > /sys/class/vfm/map
```
Hope this helps.

**Comment by antoyo** [comment]
@cgutman: I continued my investigation and it seems you stumbled on both of these issues in the past that I'm currently facing.

The first one is that I only see the first frame (similar to what happened to you [in this comment](https://github.com/moonlight-stream/moonlight-embedded/issues/573#issuecomment-1278281837)).
I added calls to `codec_get_vdec_state` like you did and I got logs like these:
```
[…]
Error: 0, status: 1 | 0x0x60
Error: 0, status: 1 | 0x0x60
Error: 0, status: 53 | 1280x720x60
Error: 0, status: 53 | 1280x720x60
Error: 0, status: 53 | 1280x720x60
Error: 0, status: 53 | 1280x720x60
Error: 0, status: 53 | 1280x720x60
Error: 0, status: 5 | 1280x720x60
[…]
```
I don't know what those statuses mean, but I only see those (1, 5 and 53) in this case.

Also, when using h264, I got the same issue that you got [in this comment](https://github.com/moonlight-stream/moonlight-embedded/issues/573#issuecomment-1336515510). Here's what I see in `dmesg`:
```
[Sat Apr 20 11:31:07 2024] 0: frame number gap error
[Sat Apr 20 11:31:08 2024] 0: frame number gap error
[Sat Apr 20 11:31:08 2024] 0: frame number gap error
[…]
```
This time, `codec_get_vdec_state` shows different statuses (6 and 22) and the error count starts increasing at some point:
```
[…]
Error: 0, status: 6 | 1280x720x60
Error: 0, status: 6 | 1280x720x60
Error: 0, status: 6 | 1280x720x60
Error: 0, status: 6 | 1280x720x60
Error: 0, status: 6 | 1280x720x60
Error: 0, status: 6 | 1280x720x60
Error: 0, status: 6 | 1280x720x60
Error: 1, status: 6 | 1280x720x60
Error: 1, status: 6 | 1280x720x60
Error: 1, status: 6 | 1280x720x60
Error: 1, status: 6 | 1280x720x60
Error: 1, status: 6 | 1280x720x60
Error: 1, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 2, status: 6 | 1280x720x60
Error: 3, status: 6 | 1280x720x60
Error: 3, status: 6 | 1280x720x60
Error: 3, status: 6 | 1280x720x60
[…]
```

I have a few questions for you:
 * How did you fix these issues (only first frame shown, "frame number gap error")?
 * What is the meaning of those status code?

Thanks for your help!

**Comment by antoyo** [comment]
Ok, I was finally able to stream some video!

I had to remove this line in `src/platform.c`:
```
write_bool("/sys/class/graphics/fb1/blank", true);
```

In order to write a proper patch, I would need to know why this was added back in https://github.com/antoyo/moonlight-embedded/commit/c2036acca10501b5f9fdcfc50e7e5001989beee1#diff-95d130942a7b36a52939da0fea295a6587be8b1249c5b7d2ac9e7018b77f561bR81.
Do you know if it would be OK to completely remove this line?
Or would I need to add a condition to only do this on the Vero V?

Thanks for your help.

**Comment by Diaoul** [comment]
@maxanier curious where do you put this start script?
Any chance to avoid the stop/start of the mediacenter and make this a separate unit that mediacenter depends on?

**Comment by TheHacker66** [comment]
Hi. I also tried again now and after uncommenting the line mentioned by
@antoyo streaming works fine @ 1080p. I get some lag @ 4k but it might be
my lan speed or something else.
I have to stop Kodi as soon as I reboot otherwise I get no video at all.
This is great news! I'm also testing from a Vero V and others in the osmc
team tested with a Vero 4k. I use Nvidia while others had to add this to
aml.c to make video work with Intel cards:

static int framecount = 0;

If (++framecount % 60 == 0) {

  return DR_NEED_IDR;

}

Just before the return at the end of the function aml_submit_decode_unit()

Thanks to everyone for testing and keeping this issue alive! And of course
thanks to moonlight devs!

Il ven 28 giu 2024, 14:46 Max Becker ***@***.***> ha scritto:

**Comment by agboom** [comment]
Hi, I'm also tinkering with my Vero V to get Moonlight to work. Thanks to this thread I was able to get it working by removing the line mentioned in https://github.com/moonlight-stream/moonlight-embedded/issues/573#issuecomment-2104492454 and building moonlight-embedded myself. 

Moonlight Embedded 274d3db ([self built](https://github.com/moonlight-stream/moonlight-embedded/wiki/Full-details-for-Compilation-and-Installation-in-Raspbian-Or-OSMC))
Sunshine: 0.23.1
OSMC: Linux vero 4.9.269-64-osmc
GPU: AMD Radeon RX 6700 XT

I'm still struggling with the optimal settings though. Setting `-bitrate 200000` cause a lot of log lines like `Unrecoverable frame 3002 (block 2 of 3): 79+10=89 received < 120 needed` and stuttering. I went all the way down to `75000` to reduce the stutters, but then had noticable lag. Not sure if those two are related.

I found a recommendation to enable performance stats with Ctrl+Alt+Shift+S, but apparently this doesn't work in moonlight-embedded (someone mentioned this in Discord). This makes it a bit more challenging to find the optimal settings. Otherwise this looks promising!

**Comment by agboom** [comment]
I continued testing Moonlight with my Vero V and it works fine, except when a black screen happens. This is resolved by rebooting the Vero. At first this seemed random, but then I noticed a pattern that every time I watched a video on OSMC, the black screen happened again. Maybe the video player does something with the frame buffer that stops Moonlight from being able to use it? 

I was able to capture the logs for both situations. There are some noticable differences, but I'm afraid I'm not knowledgable to make any sense of it.

With video:
- [dmesg output](https://paste.osmc.tv/yumejifive.sm)
- [moonlight verbose logs](https://paste.osmc.tv/evulitixuy.vhdl)

Without video:
- [dmesg output](https://paste.osmc.tv/yunatubani.sm)
- [moonlight verbose logs](https://paste.osmc.tv/vixasufowa.erl)

**Comment by antoyo** [comment]
I noticed as well that playing a video in OSMC changes some settings.

This might help you: I run moonlight-embedded from Luna (I believe I use [this add-on](https://github.com/TheChoconut/Luna)) and I modified the `prescript.sh` to the following:
```
echo 1 > /sys/class/graphics/fb0/blank
echo "rm default" > /sys/class/vfm/map
echo add default decoder ppmgr deinterlace amlvideo amvideo > /sys/class/vfm/map
```
Perhaps you need to run some of these commands before running moonlight.

**Comment by agboom** [comment]
Hi @antoyo, thanks for replying! Indeed, I use Luna as well and shared some possible improvements. Thanks for your suggestion, I'll try it as soon as I have the time.

If you have more changes you want to share, I'm certainly interested. It would be nice to have a plug-and-play experience with Moonlight for Kodi/Vero and Luna seems like the way to go right now. I have some more ideas for the long term, like app shortcuts in the main menu and automatic wake-on-lan.

**Comment by agboom** [comment]
@antoyo Your solution works, thanks! Interestingly, the first line looks like it does the same as the (removed) line from `src/platform.c`:
```c
write_bool("/sys/class/graphics/fb1/blank", true);
```
Do the commands after resolve something maybe? The vfm map seems to be accessible from MBE so maybe we can add these lines to `src/video/aml.c` or something? I'll try this later.

One issue solved, another pops up :see_no_evil:: 
The video seems to be increasingly delayed. In one session - after 30 minutes - the delay was about 2 seconds (timed from control press to video feedback). It doesn't seem input lag, because the audio feedback is almost instant. A new session resets the delay, but it's reproducible.

I have a moonlight log from the session the delay occurs: https://paste.osmc.tv/elogetokay.coffee

**Comment by antoyo** [comment]
> @antoyo Your solution works, thanks! Interestingly, the first line looks like it does the same as the (removed) line from `src/platform.c`:
> 
> ```c
> write_bool("/sys/class/graphics/fb1/blank", true);
> ```

This is setting fb1 while the command in the script sets fb0. I guess it might work to change moonlight-embedded to use fb0, I don't know.

> One issue solved, another pops up 🙈: The video seems to be increasingly delayed. In one session - after 30 minutes - the delay was about 2 seconds (timed from control press to video feedback). It doesn't seem input lag, because the audio feedback is almost instant. A new session resets the delay, but it's reproducible.
> 
> I have a moonlight log from the session the delay occurs: https://paste.osmc.tv/elogetokay.coffee

Is this caused by the commands I posted? I believe you'll find other people that had delays in this thread: perhaps they also posted the solution?
I also remember having delay at some point: I don't remember exactly how I fixed it (perhaps it was fixed by changing the codec), but here's my moonlight config in case it helps:

```ini
[Moonlight]
address = myhostname
width = 1920
height = 1080
fps = 30
bitrate = 20000
packetsize = 1392
sops = false
remote = auto
localaudio = false
debug = false
audio = sysdefault
codec = h265
nounsupported = false
nomouseemulation = true
```

**Comment by agboom** [comment]
> This is setting fb1 while the command in the script sets fb0. I guess it might work to change moonlight-embedded to use fb0, I don't know.

Ah, good point, I missed that.

> Is this caused by the commands I posted? I believe you'll find other people that had delays in this thread: perhaps they also posted the solution?

You're right, this has been mentioned before. I have `fps` at `60`, so that probably causes the delays. I should read the comments in here more closely :sweat_smile:

**Comment by zjoasan** [comment]
From TheHacker666:

> static int framecount = 0; 
> If (++framecount % 60 == 0) {
>      return DR_NEED_IDR;
> }

I had this problem to but on steriods, 30 secs streaming,15 sec frozen ideo still  audio, then rinse and repeat the isuue was that each "blackout became longer and longer. With help of another dev  at osmc.We xperrimented with clearing the buffer at intervals..

Why I probably had it worse then most is my old desktop, Intel HD 4600, first gen with some sort of quicksync. I'm guessing it's giving of a much "dirtier" stream since have i harder time keeping p with the system.

The thought was force "clearing" Veros buffer, in order to reduce the build up causing lag. Tjhe "modula 60" was to have it as a starting point. I was going to experiment with it, but I've gotten tied up in other projects. As as you have it eveny divideable by the frame rate, it "should be good".

**Comment by agboom** [comment]
@zjoasan Thanks for your suggestion. I managed to add this patch. Sadly for me it doesn't help with the increasing latency. I probably need to find the right modulo value. I set it to 60 while keeping the fps at 60 too.

Another side effect of this patch is that there's some kind of artifact pulsing about every second. If I change the modulo to 10 it pulsates faster (looks like about 6 per second), but the latency still increases over time.

I tried to record a video of it, I hope this gives an idea:

https://github.com/user-attachments/assets/193ba4b5-6515-4d6a-bd02-40edc09aaee8

**Comment by agboom** [comment]
After removing the `framecount` patch I fiddled a bit more with the Moonlight settings:
- 1080p / 60fps / 20mbps: results in increasing video latency fairly quickly (5 minutes)
- 1080p / 60fps / 10mbps:  same
- 1080p / 30fps / 10mbps: does not give noticeable latency 
- 1080p / 30fps / 20mbps: no latency as well

The behavior seems to be the same as described here: https://github.com/moonlight-stream/moonlight-embedded/issues/662

**Comment by cgutman** [comment]
This is completely unrelated to #662. In the case of #662, the Pi 2 was not rated for 1080p60, so it could not keep up without overclocking. The issue here is the same accumulation of frames due to mismatching between the rate of display and the rate of incoming frames problem I reported a long time ago: https://github.com/moonlight-stream/moonlight-embedded/issues/573#issuecomment-1308253584

We're reporting PTS values for every frame we submit, so I don't know what else we're supposed to do. We need some way of asking for overdue frames in the pipeline to be dropped or to at least limit the number of buffered frames. I'm not sure if there's a way to do that.

TBH though, I think the answer is for the Amlogic software stack to leave this poorly-documented vendor-specific API and ancient kernel version in the past and start supporting standard V4L2, so you can use a modern client like [moonlight-qt](https://github.com/moonlight-stream/moonlight-qt). Moonlight-Embedded is a legacy client from the dark ages when every SoC vendor invented their own bespoke video decoding and rendering APIs. Fortunately we're (mostly) past that era, and Moonlight-Qt's embedded packages already work great using standard V4L2+DRM+GL APIs on plenty of platforms like RK3288, RK3399, JH-7110, BCM2711, BCM2712, and probably more.

**Comment by zjoasan** [comment]
I totally agree with @cgutman , there are better solutions, but as we are tied to this platform for now. I'm still trying to get it as good as it can get. With my old hardware, a modulo of 20 eliminate build up, at least for 20-ish minutes (60 fps). There is side effects, but hard to describe. 

Thanks for the info about PTS, will take it back to OSMC-devs and see if we can come up with more elegant solution.

**Comment by agboom** [comment]
I agree, I didn't mean to imply that the issue was related, just that the behavior was the same. But yeah, RPI is a completely different thing.

Having a working moonlight-qt would be awesome. I did try it out, but got no further than the main screen, no streaming. That makes sense now based on what you explained. I gathered at some point it had something to do with DRM and KMS, or lack thereof, and a newer kernel version would help, but my knowledge is very limited in that area. It's also out of scope of this issue, so I'll stop talking about it here. I have a [thread over at OSMC Discourse](https://discourse.osmc.tv/t/dependencies-for-moonlight/107842) about Moonlight dependencies if you'd like to weigh in, I certainly would like to learn!

Anyway, regarding MBE, I've settled on 30fps for daily use. I'll try some more modulo values later. If I can help by testing anything, please let me know.
